portable/GCC/ARM_CRx_MPU/port.c

```diff
@@ -0,0 +1,799 @@
+ * @param xRegions The MPU settings being requested by the task.
+ * @param pxBottomOfStack The base address of the Task's Stack
+ * @param ulStackDepth The length of the task's stack.
+ *
+ * @ingroup Task Context
+ * @ingroup MPU Control
+ *
+ * @note pxBottomOfStack must be aligned to a region size of length ulStackDepth.
+ * @note ulStackDepth must be a power of 2 larger than 32 bytes.
+ */
+/* PRIVILEGED_FUNCTION */ void vPortStoreTaskMPUSettings(
+    xMPU_SETTINGS * xMPUSettings,
+    const struct xMEMORY_REGION * const xRegions,
+    StackType_t * pxBottomOfStack,
+    uint32_t ulStackDepth
+)
+{
+#if defined( __ARMCC_VERSION )
+
+    /* Declaration when these variable are defined in code instead of being
+     * exported from linker scripts. */
+    extern uint32_t * __SRAM_segment_start__;
+    extern uint32_t * __SRAM_segment_end__;
+    extern uint32_t * __privileged_data_start__;
+    extern uint32_t * __privileged_data_end__;
+#else
+    /* Declaration when these variable are exported from linker scripts. */
+    extern uint32_t __SRAM_segment_start__[];
+    extern uint32_t __SRAM_segment_end__[];
+
+#endif /* if defined( __ARMCC_VERSION ) */
+    uint32_t lIndex = 0x0;
+    uint32_t ul = 0x0;
+
+    if( xRegions == NULL )
+    {
+        /* No MPU regions are specified so allow access to all of the RAM. */
+        xMPUSettings->xRegion[ 0 ].ulRegionBaseAddress = ( uint32_t
+        ) __SRAM_segment_start__;
+        xMPUSettings->xRegion[ 0 ].ulRegionSize = ( prvGetMPURegionSizeSetting(
+                                                      ( uint32_t ) __SRAM_segment_end__ -
+                                                      ( uint32_t ) __SRAM_segment_start__
+                                                  ) ) |
+                                                  portMPU_REGION_ENABLE;
+        xMPUSettings->xRegion[ 0 ].ulRegionAttribute = portMPU_PRIV_RW_USER_RW_NOEXEC |
+                                                       portMPU_NORMAL_OIWTNOWA_SHARED;
+
+        /* Invalidate all other regions. */
+        for( ul = 1; ul <= portNUM_CONFIGURABLE_REGIONS; ul++ )
+        {
+            xMPUSettings->xRegion[ ul ].ulRegionBaseAddress = 0x0UL;
+            xMPUSettings->xRegion[ ul ].ulRegionSize = 0x0UL;
+            xMPUSettings->xRegion[ ul ].ulRegionAttribute = 0x0UL;
+        }
+    }
+    else
+    {
+        /* This function is called automatically when the task is created - in
+         * which case the stack region parameters will be valid. At all other
+         * times the stack parameters will not be valid and it is assumed that the
+         * stack region has already been configured. */
+
+        if( ulStackDepth > 0 )
+        {
+            uint32_t ulSmallestRegion = prvGetMPURegionSizeSetting( ulStackDepth * 0x4 );
+            /* Define the region that allows access to the stack. */
+            xMPUSettings->xRegion[ 0 ].ulRegionBaseAddress = ( uint32_t ) pxBottomOfStack;
```

Comment:
It looks like there is a mismatch where xRegion[ 0 ] is being used as the stack MPU region where portmacro_asm.h states this should be the first user configurable region. This region ordering could allow the per task user defined regions to override stack protections.

Reply:
Region number fixed, thank you for catching this